### DISCUSSION TITLE
Strange error when trying to run the AI-deck classification demo

### DISCUSSION THREAD

**Discussion by Kiesmith** [bitcraze/discussions#555] [Q&A] [answered]
I am using the AI-deck 1.1 REVD.
I am receiving a strange error at the last part of the classification demo [https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/ai-examples/classification-demo/](url)

When I run the command: **$ docker run --rm -v ${PWD}:/module --privileged aideck-with-autotiler tools/build/make-example examples/ai/classification all run**

I receive this in the window after:

bitcraze@ubuntu:~/aideck-gap8-examples$ docker run --rm -v ${PWD}:/module --privileged aideck-with-autotiler tools/build/make-example examples/ai/classification all run
Collecting numpy==1.22.3
  Using cached numpy-1.22.3-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)
ERROR: imageio 2.17.0 has requirement pillow>=8.3.2, but you'll have pillow 7.2.0 which is incompatible.
Installing collected packages: numpy
  Attempting uninstall: numpy
    Found existing installation: numpy 1.19.5
    Uninstalling numpy-1.19.5:
      Successfully uninstalled numpy-1.19.5
Successfully installed numpy-1.22.3
Building GAP8 mode with 8 bit quantization
script model/nntool_script
GEN ... /gap_sdk/tools/autotiler_v3/CNN_Generators/CNN_Generator_Util.c /gap_sdk/tools/autotiler_v3/CNN_Generators/CNN_Copy_Generators.c /gap_sdk/tools/autotiler_v3/CNN_Generators/SSD_Generators.c /gap_sdk/tools/autotiler_v3/Generators/BilinearResizes/ResizeGenerator.c  /gap_sdk/tools/autotiler_v3/DSP_Generators/DSP_Generators.c /gap_sdk/tools/autotiler_v3/CNN_Generators_SQ8/CNN_Generators_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Generators_SQ8/RNN_Generators_SQ8.c
APP_SRCS... classification.c ../../../lib/cpx/src/com.c ../../../lib/cpx/src/cpx.c BUILD_MODEL_SQ8BIT/classificationKernels.c /gap_sdk/libs/gap_lib/img_io/ImgIO.c /gap_sdk/tools/autotiler_v3/CNN_Libraries/SSD_BasicKernels.c /gap_sdk/tools/autotiler_v3/Generators/BilinearResizes/ResizeBasicKernels.c /gap_sdk/tools/autotiler_v3/CNN_Libraries/CNN_Copy.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_AT_Misc.c /gap_sdk/tools/autotiler_v3/DSP_Libraries/CmplxFunctions.c /gap_sdk/tools/autotiler_v3/DSP_Libraries/FFT_Library.c /gap_sdk/tools/autotiler_v3/DSP_Libraries/MfccBasicKernels.c /gap_sdk/tools/autotiler_v3/DSP_Libraries/PreProcessing.c /gap_sdk/tools/autotiler_v3/DSP_Libraries/math_funcs.c  /gap_sdk/tools/autotiler_v3/DSP_Libraries/pulp_dsp/plp_cos_f32s_xpulpv2.c /gap_sdk/tools/autotiler_v3/DSP_Libraries/pulp_dsp/plp_sin_f32s_xpulpv2.c  /gap_sdk/tools/autotiler_v3/DSP_Libraries/pulp_dsp/plp_common_tables.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_Activation_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_Activation_HWC_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_Bias_Linear_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_Conv_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_MatMul_Conv_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_Pooling_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_Conv_DW_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_MatAlgebra_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/CNN_SoftMax_SQ8.c /gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8/RNN_SQ8.c
APP_CFLAGS... -DMODEL_QUANTIZED -g -Os -mno-memcpy -fno-tree-loop-distribute-patterns -I. -I/gap_sdk/libs/gap_lib/include -I/gap_sdk/tools/autotiler_v3/Emulation -I/gap_sdk/tools/autotiler_v3/Autotiler -I/gap_sdk/tools/autotiler_v3/Generators/BilinearResizes -I/gap_sdk/tools/autotiler_v3/CNN_Libraries -I/gap_sdk/tools/autotiler_v3/DSP_Libraries -I/gap_sdk/tools/autotiler_v3/CNN_Libraries_fp16 -I/gap_sdk/tools/autotiler_v3/CNN_Libraries_SQ8 -I/module/examples/ai/classification/BUILD_MODEL_SQ8BIT -DPERF -DAT_MODEL_PREFIX=classification -DAT_INPUT_HEIGHT= -DAT_INPUT_WIDTH= -DAT_INPUT_COLORS= -DSTACK_SIZE=6096 -DSLAVE_STACK_SIZE=1024 -DconfigUSE_TIMERS=1 -DINCLUDE_xTimerPendFunctionCall=1 -DFS_PARTITIONTABLE_OFFSET=0x40000 -DFREQ_FC=200 -DFREQ_CL=170 -DTXQ_SIZE=5 -DRXQ_SIZE=5
gapy --target=ai_deck --platform=board --work-dir=/module/examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS --config-ini=/module/examples/ai/classification/config.ini  --config-opt=**/flash/content/partitions/readfs/files=/module/examples/ai/classification/BUILD_MODEL_SQ8BIT/classification_L3_Flash_Const.dat   run --image --binary=/module/examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/classification
gapy --target=ai_deck --platform=board --work-dir=/module/examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS --config-ini=/module/examples/ai/classification/config.ini  --config-opt=**/flash/content/partitions/readfs/files=/module/examples/ai/classification/BUILD_MODEL_SQ8BIT/classification_L3_Flash_Const.dat   run --flash --binary=/module/examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/classification
Open On-Chip Debugger 0.10.0+dev-gb84d97ec4 (2022-04-20-14:26)
Licensed under GNU GPL v2
For bug reports, read
http://openocd.org/doc/doxygen/bugs.html
debug_level: 0

DEPRECATED! use 'adapter speed' not 'adapter_khz'
DEPRECATED! use 'adapter driver' not 'interface'
TARGET create
GAP8 INIT TARGET
Error: no device found
Error: unable to open ftdi device with vid 15ba, pid 002a, description 'Olimex OpenOCD JTAG ARM-USB-TINY-H', serial '*' at bus location '*'

Flashing image with command:
gap8-openocd -d0 -c "gdb_port disabled; telnet_port disabled; tcl_port disabled" -c "script interface/ftdi/olimex-arm-usb-tiny-h.cfg; script target/gap8revb.tcl; script tcl/flash_image.tcl; script tcl/jtag_boot.tcl; gap_flash_raw_hyper /module/examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img 747792 /gap_sdk/utils/openocd_tools; exit;"
Traceback (most recent call last):
  File "/gap_sdk/utils/gapy/gapy", line 162, in <module>
    main()
  File "/gap_sdk/utils/gapy/gapy", line 152, in main
    operationFunc(args, config, system)
  File "/gap_sdk/utils/gapy/run.py", line 184, in operationFunc
    raise RuntimeError('Runner has failed with value: %d' % status)
RuntimeError: Runner has failed with value: -1
make: *** [/gap_sdk/rtos/freeRTOS/vendors/gwt/rules/freeRTOS_rules.mk:407: flash_fs] Error 1

**Note:** 
Everything seems to work just fine before, I have successfully run the wifi stream and I managed to run the GAP8 led fast blinking too. 
I have tried updating pillow since the log says there is a problem with the one I use. I cant uninstall it and when I install one, it keeps using the same pillow 7.0.

It would be really appreciated if I can get help on this as I have spent allot of hours already trying to solve it. 
Thankyou and best regards,
Kie

**Comment by hmllr** [comment, marked as answer]
Hi :)
I am not sure if the error message about the pillow version is crucial - but let's fix the other error message about the debugger not being found first.
"Error: no device found
Error: unable to open ftdi device with vid 15ba, pid 002a, description 'Olimex OpenOCD JTAG ARM-USB-TINY-H', serial '' at bus location ''"
Is important in this error log, it means that your debugger is not found. Do you have an Olimex OpenOCD JTAG ARM-USB-TINY-H connected to your computer? (you can check this with "lsusb" in a terminal). And did you set the udev permissions for it?
Hope this helps,
Hanna

**Reply by Kiesmith** [reply]
I do not believe I have what you are talking about. I was only given the AI deck and the only thing I use to connect to it is the bitcraze crazyradio PA USB dongle (as seen in the image).

![image](https://user-images.githubusercontent.com/72498028/217388947-c0d0493a-dc6a-44c4-894a-9f3af87c050e.png)

I never heard of the Olimex OpenOCD JTAG ARM-USB-TINY-H and I had to google it. If you're talking about these: 
![image](https://user-images.githubusercontent.com/72498028/217389355-ab2d1607-8c95-49fe-b0aa-dcb9c1ac4f6a.png)

I do not have one. 


Also I just tried the command you gave me (if this helps): 
![image](https://user-images.githubusercontent.com/72498028/217391418-916afb90-dbe9-4a58-8a4e-168a75855f62.png)

**Reply by hmllr** [reply]
ok, thanks for the detailed description!

There are two possible ways to flash the AI-deck, either over JTAG with a debugger (as the olimex), so over a physical connection directly to the chip or using the cfloader over-the-air (wirelessly transmitting the code you want to flash to the Crazyflie), using the bootloader which is already flashed on your AI-deck as you have a revD. As you do not have a debugger, you have to  use the cfloader (which is anyway the recommended option). Those methods are explained here: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/getting-started/flashing/ 

Now, if you follow the classification example documentation then you should follow the commands in "Execute the image classification CNN on the AI-deck" (https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/ai-examples/classification-demo/#execute-the-image-classification-cnn-on-the-ai-deck)  which do work without any debugger connected, you only need the Crazyradio plugged in and your Crazyflie turned on (with all firmwares up-to-date).

Hope this helps,
Hanna

**Reply by Kiesmith** [reply]
Do you know how I would write this one in specific over the radio? I tried doing it the same way as the wifi example but that seems to flash a .bin over. 

**Update**

So I believe I have done this with reference to another post. I input: 

**$ cd aideck-gap8-examples/examples/ai/classification/BUILD/GAP8_V2/
GCC_RISCV_FREERTOS$**

**$ cfloader flash ./target.board.devices.flash.img deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E7E7**

This uploaded whatever I sent to the drone which is okay. The classification guide on the website then said I should expect to see a screen showing the AI detecting and not detecting things. I do not see that. I saw that some people are using the **opencv_viewer.py** but I am a bit confused how to open it since this is part of the **WiFi Video Streamer** guide.

How can I access the video to verify the AI is working if everything I have done previously is correct? 
If I do the Wi-Fi video streamer guide, wouldn't this over-write the ai program I just flashed over? 

Thankyou for your help so far also 😃

**Reply by hmllr** [reply]
what do you mean by "this one in specific"?
If you mean the classifier example, it tells you to execute: `cfloader flash examples/ai/classification/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img deck-bcAI:gap8-fw -w [CRAZYFLIE URI]` in the docs.
The wifi example from the getting started guide is a bit special, because there you don't build the project yourself but download the binary. The .bin there and the target.board.devices.flash.img that gets generated when you build the wifi example are equivalent.
Hope this helps,
Hanna

**Reply by Kiesmith** [reply]
It has flashed over to the AI and I am just checking the client console. I can see the feed being updated, I am just unsure how to access the video. 

How do I access the camera footage like in the picture? 

This is what I am referring to: 
![image](https://user-images.githubusercontent.com/72498028/217654280-87b73d41-c002-43c5-b268-21e148c18517.png)

**Reply by hmllr** [reply]
the gif is an outside view of the scene you should see. You do not see the video streaming in the classification example. This is only to show you the output on the terminal while you see if there is a parcel placed in the FoV or not.
Note that you don't get the output in the terminal if you don't run over JTAG (however you get it into the cfclient console).
If you want the video from the camera streamed you need to combine the code from the video streamer with the classification example yourself.
Hope this clarifies things,
Hanna

**Reply by Kiesmith** [reply]
Thankyou very much for the help on this!
